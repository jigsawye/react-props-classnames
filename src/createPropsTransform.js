import React from 'react';
import PropTypes from 'prop-types';

import PropsTransform from './PropsTransform';

const getDisplayName = WrappedComponent =>
  WrappedComponent.displayName || WrappedComponent.name || 'Unknown';

export default options => Component => {
  const TransformedComponent = ({ className, children, ...otherOrops }) => {
    const propsTransform = new PropsTransform(options);
    const propsClassName = propsTransform.getClassNameFromProps(otherOrops);
    const mergedClassName = `${className} ${propsClassName}`.trim();

    return (
      <Component className={mergedClassName} {...otherOrops}>
        {children}
      </Component>
    );
  };

  TransformedComponent.displayName = `PropsTransformer(${getDisplayName(
    Component
  )})`;

  TransformedComponent.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
  };

  TransformedComponent.defaultProps = {
    children: '',
    className: '',
  };

  return TransformedComponent;
};
