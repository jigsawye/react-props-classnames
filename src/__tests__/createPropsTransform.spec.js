import React from 'react';
import { render, mount } from 'enzyme';
import styled from 'styled-components';

import createPropsTransform from '../createPropsTransform';

describe('createPropsTransform', () => {
  it('should export createPropsTransform', () => {
    expect(createPropsTransform).toBeDefined();
  });

  it('should create propsTansform', () => {
    const propsTansform = createPropsTransform();

    expect(propsTansform).toBeDefined();
  });

  it('should transform component props', () => {
    const propsTansform = createPropsTransform();

    const StyledComp = styled.div`
      background-color: #fff;
    `;
    const TransformedComponent = propsTansform(StyledComp);
    const component = render(<TransformedComponent disabled size="lg" />);

    expect(component).toMatchSnapshot();
    expect(component.prop('class')).toMatch(/default-prefix-disabled/);
    expect(component.prop('class')).toMatch(/default-prefix-size-lg/);
  });

  it('should transform component props with specified `props`', () => {
    const propsTansform = createPropsTransform({
      prefix: 'test',
      props: ['disabled', 'size'],
    });

    const StyledComp = styled.div`
      background-color: #fff;
    `;
    const TransformedComponent = propsTansform(StyledComp);
    const component = render(
      <TransformedComponent disabled size="lg" light type="primary" />
    );

    expect(component).toMatchSnapshot();
    expect(component.prop('class')).toMatch(/test-disabled/);
    expect(component.prop('class')).toMatch(/test-size-lg/);
    expect(component.prop('class')).not.toMatch(/test-light/);
    expect(component.prop('class')).not.toMatch(/test-type-primary/);
  });

  it('should render component name as `PropsTransformer(DisplayName)`', () => {
    const propsTansform = createPropsTransform({
      prefix: 'test',
      props: ['disabled', 'size'],
    });

    const StyledComp = styled.div`
      background-color: #fff;
    `;
    const TransformedStyleComp = propsTansform(StyledComp);
    const StyledComponent = mount(<TransformedStyleComp />);

    const FunctionalComp = () => <div />;
    const TransformedFunctionalComp = propsTansform(FunctionalComp);
    const FunctionalComponent = mount(<TransformedFunctionalComp />);

    const UnknownComp = () => () => <div />;
    const TransformedUnknownComp = propsTansform(UnknownComp());
    const UnknownComponent = mount(<TransformedUnknownComp />);

    expect(StyledComponent).toMatchSnapshot();
    expect(StyledComponent.name()).toEqual('PropsTransformer(styled.div)');

    expect(FunctionalComponent).toMatchSnapshot();
    expect(FunctionalComponent.name()).toEqual(
      'PropsTransformer(FunctionalComp)'
    );

    expect(UnknownComponent).toMatchSnapshot();
    expect(UnknownComponent.name()).toEqual('PropsTransformer(Unknown)');
  });
});
