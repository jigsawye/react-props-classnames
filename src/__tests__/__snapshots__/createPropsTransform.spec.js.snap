// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPropsTransform should render component name as \`PropsTransformer(DisplayName)\` 1`] = `
.c0 {
  background-color: #fff;
}

<PropsTransformer(styled.div)
  className=""
>
  <styled.div
    className=""
  >
    <div
      className="c0"
    />
  </styled.div>
</PropsTransformer(styled.div)>
`;

exports[`createPropsTransform should render component name as \`PropsTransformer(DisplayName)\` 2`] = `
<PropsTransformer(FunctionalComp)
  className=""
>
  <FunctionalComp
    className=""
  >
    <div />
  </FunctionalComp>
</PropsTransformer(FunctionalComp)>
`;

exports[`createPropsTransform should render component name as \`PropsTransformer(DisplayName)\` 3`] = `
<PropsTransformer(Unknown)
  className=""
>
  <Component
    className=""
  >
    <div />
  </Component>
</PropsTransformer(Unknown)>
`;

exports[`createPropsTransform should transform component props 1`] = `
.c0 {
  background-color: #fff;
}

<div
  class="default-prefix-disabled default-prefix-size-lg c0"
  disabled=""
/>
`;

exports[`createPropsTransform should transform component props with specified \`props\` 1`] = `
.c0 {
  background-color: #fff;
}

<div
  class="test-disabled test-size-lg c0"
  disabled=""
  type="primary"
/>
`;
